manager/src/main/java/com/redhat/service/bridge/manager/connectors/AbstractConnectorWorker.java

```diff
@@ -0,0 +1,126 @@
+package com.redhat.service.bridge.manager.connectors;
+
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.UUID;
+
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.transaction.Transactional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.redhat.service.bridge.manager.models.ConnectorEntity;
+
+import io.vertx.mutiny.core.eventbus.EventBus;
+
+// R is the type returned by the external service
+public abstract class AbstractConnectorWorker<R> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractConnectorWorker.class);
+
+    @Inject
+    protected EntityManager entityManager;
+
+    @Inject
+    protected EventBus eventBus;
+
+    protected abstract R callExternalService(ConnectorEntity connectorEntity);
+
+    // Don't need to implement this if this is a deletion worker
+    protected ConnectorEntity updateEntityForSuccess(ConnectorEntity connectorEntity, R serviceResponse) {
+        return connectorEntity;
+    }
+
+    protected abstract ConnectorEntity updateEntityForError(ConnectorEntity connectorEntity, Throwable error);
+
+    protected void afterSuccessfullyUpdated(ConnectorEntity c) {
+    }
+
+    // Used for workers that should delete the entity after successfully calling the service.
+    // By default, after the service call updateEntityForSuccess is called, with this physical deletion will occur
+    protected boolean shouldDeleteAfterSuccess() {
+        return false;
+    }
+
+    // Could be the k8s pod name
+    String workerId = UUID.randomUUID().toString();
+
+    @Transactional(Transactional.TxType.NEVER)
+    public void execute(ConnectorEntity currentEntity) {
+
+        ConnectorEntity connectorEntity;
+        try {
+            connectorEntity = claimConnector(currentEntity);
+            LOGGER.info("Running worker: {} on Entity: {} with workerId: {} ", workerName(), connectorEntity, workerId);
+        } catch (Exception e) {
+            // Cannot claim, return the unmodified entity
+            LOGGER.trace("Worker {} cannot claim Entity: {} with error: {} ", workerId, workerName(), e);
+            return;
+        }
+
+        try {
+            R serviceResponse = callExternalService(connectorEntity);
+
+            if (!shouldDeleteAfterSuccess()) {
+
+                ConnectorEntity successConnectorEntity = successfullyCalledService(connectorEntity, serviceResponse);
+                afterSuccessfullyUpdated(successConnectorEntity);
+            } else {
+                deleteEntity(connectorEntity);
+            }
+        } catch (Exception e) {
+
+            errorWhileCalling(e, connectorEntity);
+        }
+    }
+
+    @Transactional(Transactional.TxType.REQUIRES_NEW)
+    public ConnectorEntity claimConnector(ConnectorEntity currentEntity) {
```

Comment:
I think there is a concurrency here. Think about the following scenario: 
1) Pod A runs `ConnectorsOrchestrator.updatePendingConnectors` and with `ConnectorDAO.findUnprocessed` finds the connector X. It sends to its internal bus the request. 
2) Pod B does the same and finds the connector X. The request is sent to its internal bus as well. 
3) Pod A starts processing and calls this `claimConnector` which sets `mergedCE.setWorkerId(workerId);`
4) Pod B does the same and overrides with its `mergedCE.setWorkerId(workerId);`

In short, I think this `claimConnector` is not really preventing multiple pods to take ownership of the resource since it is not checking that the `workerId` is still `null`.  Or am I missing something?

Reply:
We have a `@Version` annotation on the `ConnectorEntity` to enforce optimistic locking. With that only one transaction will be able to successfully commit the claim on the entity. The other will fail because the other transaction committed first and updated the version of the Entity. 

Essentially the `@Version` annotation makes the logic of the db write something like:

```
int updatedRows = persistEntity(entity);
if(updatedRows == 0){
throw OptimisticLockingException()
}
```

The SQL generated for the update looks something like `update table connector set id=1 where version=1`

If the `version` has changed then the number of rows updated is 0 and the commit fails.

Reply:
`CONNECTORENTITY.findUnprocessed` checks for the unprocessed entities (those who don't have the workerId) and the claim is a lock that prevents two worker from working on the same entity

Reply:
Closing this as it should be cleared